awsplugins/ec2/ec2.go

```diff
@@ -9,35 +9,115 @@
 package ec2
 
 import (
-	"github.com/aws/aws-sdk-go/aws/ec2metadata"
-	"github.com/aws/aws-sdk-go/aws/session"
+	"bytes"
+	"encoding/json"
+	"net/http"
+
 	"github.com/aws/aws-xray-sdk-go/internal/logger"
 	"github.com/aws/aws-xray-sdk-go/internal/plugins"
 )
 
 // Origin is the type of AWS resource that runs your application.
 const Origin = "AWS::EC2::Instance"
 
-// Init activates EC2Plugin at runtime.
+type Metdata struct {
+	AccountID               string
+	Architecture            string
+	AvailabilityZone        string
+	BillingProducts         string
+	DevpayProductCodes      string
+	MarketplaceProductCodes string
+	ImageID                 string
+	InstanceID              string
+	InstanceType            string
+	KernelID                string
+	PendingTime             string
+	PrivateIP               string
+	RamdiskID               string
+	Region                  string
+	Version                 string
+}
+
+//Init activates EC2Plugin at runtime.
 func Init() {
 	if plugins.InstancePluginMetadata != nil && plugins.InstancePluginMetadata.EC2Metadata == nil {
 		addPluginMetadata(plugins.InstancePluginMetadata)
 	}
 }
 
 func addPluginMetadata(pluginmd *plugins.PluginMetadata) {
-	session, e := session.NewSession()
-	if e != nil {
-		logger.Errorf("Unable to create a new ec2 session: %v", e)
-		return
+	var instanceData Metdata
+	imdsURL := "http://169.254.169.254/latest/"
+
+	client := &http.Client{
+		Transport: http.DefaultTransport,
+	}
+
+	token, err := getToken(imdsURL, client)
+	if err != nil {
+		logger.Errorf("Unable to fetch EC2 instance metadata token fallback to IMDS V1: %v", err)
 	}
-	client := ec2metadata.New(session)
-	doc, err := client.GetInstanceIdentityDocument()
+
+	resp, err := getMetadata(imdsURL, client, token)
 	if err != nil {
 		logger.Errorf("Unable to read EC2 instance metadata: %v", err)
 		return
 	}
 
-	pluginmd.EC2Metadata = &plugins.EC2Metadata{InstanceID: doc.InstanceID, AvailabilityZone: doc.AvailabilityZone}
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(resp.Body); err != nil {
+		logger.Errorf("Error while reading data from response buffer: %v", err)
+	}
+	metadata := buf.String()
+
+	if err := json.Unmarshal([]byte(metadata), &instanceData); err != nil {
+		logger.Errorf("Error while unmarshal operation: %v", err)
+	}
+
+	pluginmd.EC2Metadata = &plugins.EC2Metadata{InstanceID: instanceData.InstanceID, AvailabilityZone: instanceData.AvailabilityZone}
 	pluginmd.Origin = Origin
 }
+
+// getToken fetches token to fetch EC2 metadata
+func getToken(imdsURL string, client *http.Client) (string, error) {
+	ttlHeader := "X-aws-ec2-metadata-token-ttl-seconds"
+	defaultTTL := "60"
+	tokenURL := imdsURL + "api/token"
+
+	req, _ := http.NewRequest("PUT", tokenURL, nil)
+	req.Header.Add(ttlHeader, defaultTTL)
+	resp, err := client.Do(req)
+	if err != nil {
+		return "fallback", err
```

Comment:
Do you think we could return `nil` instead of a hardcoded string when the token request fails? It seems unnecessarily complicated to me to have a specific string for an invalid token.

Reply:
Unfortunately we have to return a string. I was thinking returning an empty string but then thought this would be better than empty string as we can identify case when there's a token failure and also can be useful in unit testing.

Reply:
Ok fair enough.

Reply:
I personally find this hardcoded string to be weird too. Either we got a token or we didn;'t, modeling the latter with empty string seems equivalent even for unit tests without introducing this magic string, right?

Reply:
If you want to stick with the string at least extract a constant - but recommend empty string.